test vector semantics

marcrasi opened the pull request:
Adds a `checkVectorSemantics` method that checks the recently-added `Scalars` and `scalars` requirements.

Migrates all the existing vector semantics checks to be inside of `checkVectorSemantics` too.

marcrasi commented:
Converting to draft because we discussed some improvements in person.

marcrasi commented:
I have updated this and it's ready for review now.

dabrahams commented on Sources/SwiftFusion/Core/FixedShapeTensor.swift:
```diff
@@ -78,6 +78,10 @@ extension FixedShapeTensor {
   public static var zero: Self { .init(Tensor(zeros: Self.shape)) }
+
+  public var zeroTangentVectorInitializer: () -> Self {
+    { Self.zero }
```
Since uses of `.zero` are a red flag that I'm trying to eliminate, we should factor things differently, e.g. make a common `static var zero_` that `zero` invokes, erwhatever.

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,283 +1,206 @@
+  /// Note: This does check the semantics of the vector math operations, but only using vectors
+  /// that have the same `scalar.indices` as `self`. If `Self` supports vector operations between
+  /// instances between different `scalars.indices`, then tests should additionally call
+  /// `checkVectorMathSemantics` with examples of vectors with differing `scalars.indices`.
```
I don't like this restriction/distinction, especially in testing code where efficiency isn't a priority.  Can you explain why it's needed?

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,283 +1,206 @@
+  /// instances between different `scalars.indices`, then tests should additionally call
+  /// `checkVectorMathSemantics` with examples of vectors with differing `scalars.indices`.
+  ///
+  /// - Parameter doesNotSupportTwoScalars: pass `true` when `Self` does not have instances with
```
```suggestion
  /// - Parameter doesNotSupportTwoScalars: pass `true` when `Self` cannot have instances with
```

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,283 +1,206 @@
+    for (i, e) in zip(mutableSelf.scalars.indices, distinctScalars) { mutableSelf.scalars[i] = e }
+    XCTAssertTrue(mutableSelf.scalars.elementsEqual(distinctScalars))
+
+    // Returns a vector with the same `scalars.indices` as `self` but with the scalars replaced
+    // by the stride from `start` by `stride`.
```
```suggestion
    /// Returns a vector with the same `scalars.indices` as `self` but with the scalars replaced
    /// by the stride from `start` by `stride`.
```

marcrasi commented:
I rebased this and updated it based on the github and in-person comments.

I also added a new test for the AnyArrayBuffer Vector conformance and found a failure: #184

dabrahams commented on Sources/SwiftFusion/Core/FixedShapeTensor.swift:
```diff
@@ -70,6 +70,10 @@ extension FixedShapeTensor {
 
   public static var zero: Self { .init(Tensor(zeros: Self.shape)) }
+
+  public var zeroTangentVectorInitializer: () -> Self {
```
Why is this part of the "test vector semantics" change?  Is it a bug fix?

marcrasi replied:
hmm, this doesn't look like it's necessary and the tests still pass when I remove it. so I'll remove it.

dabrahams commented on Sources/SwiftFusion/Core/FixedShapeTensor.swift:
```diff
@@ -70,6 +70,10 @@ extension FixedShapeTensor {
 
   public static var zero: Self { .init(Tensor(zeros: Self.shape)) }
+
+  public var zeroTangentVectorInitializer: () -> Self {
```
Might as well doc comment it

dabrahams commented on Sources/SwiftFusion/Inference/ArrayBuffer+Vector.swift:
```diff
@@ -158,6 +158,7 @@ extension ArrayBuffer: Vector where Element: Vector {
   @differentiable
   private static func minus (_ lhs: ArrayBuffer, _ rhs: ArrayBuffer) -> ArrayBuffer {
     if rhs.isEmpty { return lhs }
+    if lhs.isEmpty { return .init(rhs.lazy.map { $0.zeroValue - $0 }) }
```
Nice catch.

Ugh, we have so many `zeroXXX`s flying around, I can't keep track.  We should think about how to simplify that.

marcrasi replied:
Yes :)

dabrahams commented on Tests/SwiftFusionTests/Core/FixedShapeTensorTests.swift:
```diff
@@ -47,12 +47,11 @@ class FixedShapeTensorTests: XCTestCase {
-  static var dimension: Int { 100 }
-  func testAll() {
-    runAllFixedSizeVectorTests()
+  func testVectorConformance() {
```
I'm having trouble reviewing this change; the old and new content don't have an obvious correspondence.  Maybe you could explain?

marcrasi replied:
`runAllFixedSizeVectorTests` is what ran the old equivalent of the vector semantics checks. I have completely replaced it with the new vector semantics checks, which test pretty much the same thing but are written very differently.

dabrahams commented on Tests/SwiftFusionTests/Core/FixedSizeMatrixTests.swift:
```diff
@@ -24,6 +24,18 @@ fileprivate typealias Matrix3x2 = FixedSizeMatrix<Array3<Vector2>>
+  func testVectorConformance() {
+    let s1 = (0..<6).lazy.map { Double($0) }
+    let v1 = Matrix3x2(s1)
+    v1.checkVectorSemantics(expecting: s1, writing: (6..<12).lazy.map { Double($0) })
```
I think this would be clearer if it were `checkVectorSemantics(expectingScalars:writingScalars:)`.

marcrasi replied:
Done

dabrahams commented on Tests/SwiftFusionTests/Core/VectorNTests.swift.gyb:
```diff
@@ -7,7 +7,7 @@ import SwiftFusion
 % import math
 % dims = range(1, 10)
```
```suggestion
% vectorSizes = range(1, 10)
```

dabrahams commented on Tests/SwiftFusionTests/Core/VectorNTests.swift.gyb:
```diff
@@ -7,7 +7,7 @@ import SwiftFusion
 
-class ConcreteVectorTests: XCTestCase {
+class VectorNTests: XCTestCase {
   % for dim in dims:
```
```suggestion
  % for N in vectorSizes:
```

Take or leave these suggestions, but I think this would end up being easier to read.

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,192 @@
-      XCTAssertEqual(pb(v).1, -v)
+extension Vector {
+  /// XCTests `self`'s semantic conformance to `Vector`, expecting its scalars to match
+  /// `expectedScalars`.
```
Say something about the other parameters.  `maxSupportedScalarCount` is especially mysterious.

marcrasi replied:
Done

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,192 @@
+
+    // Check that setting `scalars` actually changes it.
+    var mutableSelf = self
+    for (i, e) in zip(mutableSelf.scalars.indices, distinctScalars) { mutableSelf.scalars[i] = e }
```
The thing you're actually testing here should have been checked by `checkMutableCollectionSemantics`, I think.  The thing to test that corresponds to the comment you wrote is `mutableSelf.scalars = …`

marcrasi replied:
Changed.

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,192 @@
+
+    // MARK: - Check the `dimension` property semantics.
+
+    XCTAssertEqual(self.dimension, expectedScalars.count)
```
Makes me wonder if `dimension` should be a requirement at all.  Could just be `{ return scalars.count }`

marcrasi replied:
Yeah. I'll do this refactoring in a separate PR.

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,192 @@
+    // `zeroValue`) semantics.
+
+    /// Returns a vector with the same `scalars.indices` as `self` but with the scalars replaced
+    /// by the stride from `start` by `stride`.
```
I can't parse this doc.  The two “by”s are confusing.  Please clarify.

marcrasi replied:
Done

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,192 @@
+    // MARK: - Check unsafe buffer semantics.
+
+    self.withUnsafeBufferPointer { b in
+      XCTAssertEqual(b.count, expectedScalars.count)
+      for (actual, expected) in zip(b, expectedScalars) {
+        XCTAssertEqual(actual, expected)
+      }
```
```suggestion
      XCTAssert(b.elementsEqual, expectedScalars)
```

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,192 @@
-    XCTAssertEqual(doMutatingScalarProduct(s, v1), makeVector(from: 2, stride: 2))
+  /// XCTests the semantics of the mutating and nonmutating addition operations at `(self, other)`,
+  /// and their derivatives.
+  private func checkPlus(_ other: Self, equals expectedResult: Self) {
```
Nice factoring.

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -322,4 +222,28 @@ class MultipleVectorTests: XCTestCase {
     XCTAssertEqual(pb(Vector5(0, 0, 0, 0, 1)).1, Vector3(0, 0, 1))
   }
+
+  func testConvertToTensor() {
```
Is this really part of vector semantics testing, or could you just not find a more suitable place to put it?

marcrasi replied:
It is testing some methods that are implemented as extensions of `Vector`. Maybe changing the class name to `VectorExtensionTests` clarifies this?

dabrahams reviewed (approved):
Some small changes I'd like to see, plus answers to questions.  Overall, though, LGTM!

dabrahams commented on Tests/SwiftFusionTests/Core/VectorTests.swift:
```diff
@@ -1,292 +1,193 @@
+    /// `start + 2 * stride`, etc.
+    func strideVector(from start: Double, by stride: Double) -> Self {
+      var r = self
+      for (i, e) in zip(
```
@marcrasi FYI, we have algorithms for this kind of open-coded loop, now!

See PenguinExtensions.swift extensions on MutableCollection

dabrahams replied:
```swift
      r.scalars.assign(
        Swift.stride(from: start, to: start + Double(r.dimension) * stride, by: stride))
```
